Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+            // will trigger the upgrade.
+            upgrader()
+        }
+    }
+    
+    private func notUpgrading(context: ChannelHandlerContext, data: HTTPClientResponsePart) {
+        self.upgradeState = .upgradeFailed
+        
+        if !self.seenFirstResponse {
+            // We haven't seen the first response .end. That means we're not buffering anything, and we can
+            // just deliver data.
+            assert(self.receivedMessages.count == 0)
+            context.fireChannelRead(self.wrapInboundOut(data))
+        } else {
+            // We've seen the first request .end, so we now need to deliver the .head we got passed,
+            // as well as the .end we swallowed, and any buffered parts. While we're doing this
+            // we may be reentrantly called, which will cause us to buffer new parts. To make that safe, we
+            // must ensure we aren't holding the buffer mutably, so no for loop for us.
+            context.fireChannelRead(self.wrapInboundOut(data))
+            context.fireChannelRead(self.wrapInboundOut(.end(nil)))
+            
+            while self.receivedMessages.count > 0 {
+                let bufferedPart = self.receivedMessages.removeFirst()
+                context.fireChannelRead(bufferedPart)
+            }
+        }
+        
+        // Ok, we've delivered all the parts. We can now remove ourselves, which should happen synchronously.
+        context.pipeline.removeHandler(context: context, promise: nil)
+    }
+    
+    public func errorCaught(context: ChannelHandlerContext, error: Error) {
+        // As we are not really interested getting notified on success or failure we just pass nil as promise to
+        // reduce allocations.
+        context.close(promise: nil)
+    }
+    
+    private func removeHandler(context: ChannelHandlerContext, handler: RemovableChannelHandler?) -> EventLoopFuture<Void> {
+        if let handler = handler {
+            return context.pipeline.removeHandler(handler)
+        } else {
+            return context.eventLoop.makeSucceededFuture(())
+        }
+    }
+}
+
+extension NIOHTTPClientUpgradeHandler {
+    /// The state of the upgrade handler.
+    fileprivate enum UpgradeState {
+        /// Request not sent. This will need to be sent to initiate the upgrade.
+        case requestRequired
+
+        /// Awaiting confirmation response which will allow the upgrade to zero one or more protocols.
+        case awaitingConfirmationResponse
+        
+        /// The response head has been received. We're currently running the future chain awaiting an upgrader.
+        case awaitingUpgrader
+        
+        /// The upgrade is in process.
+        case upgrading
+        
+        /// The upgrade is in process and all of the http handlers have been removed.
+        case upgradingAddingHandlers
+        
+        /// We have an upgrader, which means we can begin upgrade.
+        case upgraderReady(() -> Void)
```

Comment:
The ordering of states here gave me a bit of confusion. It's normally a bit helpful to try to arrange the states in roughly the order in which they'd flow, in which case `upgraderReady` should come before `upgrading` and `upgradingAddingHandlers`.